src/modules/sway/bar.cpp

```diff
@@ -78,21 +132,68 @@ void BarIpcClient::onIpcEvent(const struct Ipc::ipc_response& res) {
   }
 }
 
+void BarIpcClient::onCmd(const struct Ipc::ipc_response& res) {
+  if (res.type == static_cast<uint32_t>(IPC_GET_WORKSPACES)) {
+    try {
+      auto payload = parser_.parse(res.payload);
+      for (auto& ws : payload) {
+        if (ws["urgent"].asBool()) {
+          spdlog::debug("Found workspace {} with urgency set. Stopping search.", ws["name"]);
+          // Found one workspace with urgency set, signal bar visibility
+          signal_urgency_(true);
+          return;
+        }
+      }
+      // Command to get workspaces was sent after a change in workspaces was based on "urgent",
+      // if no workspace has this flag set to true, all flags must be cleared.
+      signal_urgency_(false);
+    } catch (const std::exception& e) {
+      spdlog::error("Bar: {}", e.what());
+    }
+  }
+}
+
 void BarIpcClient::onConfigUpdate(const swaybar_config& config) {
   spdlog::info("config update for {}: id {}, mode {}, hidden_state {}", bar_.bar_id, config.id,
                config.mode, config.hidden_state);
   bar_config_ = config;
   update();
 }
 
+void BarIpcClient::onModeUpdate(bool visible_by_mode) {
+  spdlog::debug("mode update for {}: {}", bar_.bar_id, visible_by_mode);
+  visible_by_mode_ = visible_by_mode;
+  update();
+}
+
 void BarIpcClient::onVisibilityUpdate(bool visible_by_modifier) {
-  spdlog::debug("visiblity update for {}: {}", bar_.bar_id, visible_by_modifier);
+  spdlog::debug("visibility update for {}: {}", bar_.bar_id, visible_by_modifier);
   visible_by_modifier_ = visible_by_modifier;
+  if (visible_by_modifier) {
+    signal_last_modifier_(true);
+  }
+  update();
+}
+
+void BarIpcClient::onModifierPress(bool state) { last_press_modifier_ = state; }
+
+void BarIpcClient::onUrgencyUpdate(bool visible_by_urgency) {
+  spdlog::debug("urgency update for {}: {}", bar_.bar_id, visible_by_urgency);
+  visible_by_urgency_ = visible_by_urgency;
   update();
 }
 
 void BarIpcClient::update() {
-  bool visible = visible_by_modifier_;
+  bool visible = visible_by_mode_ || visible_by_modifier_ || visible_by_urgency_;
+  if (!visible_by_modifier_ && last_press_modifier_) {
+    // Modifier key was pressed and released without a different action.
+    // This signals an acknowledgment and should hide the bar again.
+    // Hide the bar and clear the mode and urgency flag.
+    visible = false;
+    visible_by_urgency_ = false;
+    visible_by_mode_ = false;
```

Comment:
That doesn't match with swaybar logic. Why do we need to track `last_press_modifier_`?
Also, you can set `visible_by_urgency_`/`visible_by_mode_` in onVisibilityUpdate.

Reply:
Thanks for feedback!

> swaybar logic - reset `visible_by_urgency_`/`visible_by_mode_` when `visible_by_modifier_` is set by the IPC event.
> 

So this is this [comment](https://github.com/swaywm/sway/blob/master/swaybar/ipc.c#L450-L452). My interpretation of this was to "ignore" `visible_by_modifier` if another reason for visibility existed, as the bar is not visible by modifier but visible by e.g. urgency already.

In my setup the modifier key that shows the bar is the same as my "normal" modifier key for all the other key bindings, this might be the reason why I implemented the logic to wait for other events after the `visible_my_modifier` to make sure it was not followed by another event, meaning that the intended action of pressing the key was not meant to clear the visibility flag.

> I think I get why you believe it is buggy - new urgent or mode event may come while the modifier is still pressed and set the flag again.

Actually I did not think about this behavior, but I think if the event comes while the key is pressed, it is still shown after releasing the key.

My observation was the following (with the sway prefix key equal the modifier key for the bar):
- if the bar is visible by urgency, pressing and releasing the mod key hides the bar: okay
- if the bar is visible by urgency, pressing the mod key in combination with an workspace change keeps the bar visible after releasing the key: according to the referenced comment not correct, but from my point of view a correct behavior
- if the bar is visible by urgency, pressing the mod key to e.g. change the focus in the same workspace hides the bar: not good
  - changing the workspace will make the bar visible again


> Maybe pinging emersion to clarify swaybar behavior would be a good idea.

If you are now also in doubt of the intended behavior it might be good the get some input, but maybe it is just that I have some edge case setup for which is was not intended and needs a fix.
Not sure if you agree on this, but if sway does not make any rules how the bar should behave, might it not be okay if different bars behave differently?

Reply:
I just confirmed that the `swaybar` behavior was copied mostly verbatim from `i3bar`. Any modifier press resets the visibility state of the bar, even if it was a part of the binding. So we can say it's intentional.

I'm fine with both options, with a slight preference towards swaybar/i3bar behavior.

> * if the bar is visible by urgency, pressing the mod key to e.g. change the focus in the same workspace hides the bar: not good
>   * changing the workspace will make the bar visible again

Actually, that second part sounds like a bug. i3 does not make the bar visible again on the workspace change, unless the last change in the `workspace` event was `"urgent"` and there are other workspaces with urgency hint set (and I can't figure out from the code why/how i3bar does that).

Reply:
I added a separate commit in which both options should be possible.
It is a bit messy, I think the bar config from the waybar config file will not be used if sway ipc is active? I guess it is not possible to see from the ipc message if a change occurs so we just use the values from sway?

One option is to reset the bar visibility right at pressing the mod key.
The other is the "release" part, which I defined as pressing the mod key and releasing it again, which for me makes more sense because it allows to start something without loosing the urgency hint and also to get rid of the bar if one intends to do so.
But I am open to feedback to adding a new config option.

I also changed the way the bar will reset visibility on a workspace change, it should now stay hidden if the urgency is cleared until a new event is received.